src/generic.jl

```diff
@@ -26,15 +26,19 @@ evaluate(dist::PreMetric, a, b) = dist(a, b)
 # Generic functions
 
 """
-    result_type(dist::PreMetric, Ta::Type, Tb::Type) -> T
-    result_type(dist::PreMetric, a::AbstractArray, b::AbstractArray) -> T
+    result_type(dist, Ta::Type, Tb::Type) -> T
+    result_type(dist, a, b) -> T
 
 Infer the result type of metric `dist` with input type `Ta` and `Tb`, or input
 data `a` and `b`.
 """
-result_type(::PreMetric, ::Type, ::Type) = Float64 # fallback
-result_type(dist::PreMetric, a::AbstractArray, b::AbstractArray) = result_type(dist, eltype(a), eltype(b))
+result_type(::PreMetric, ::Type, ::Type) = Float64 # fallback in Distances
+result_type(f, a::Type, b::Type) = typeof(f(oneunit(a), oneunit(b))) # don't require `PreMetric` subtyping
 
+# Promote Arrays and Numbers to types
+result_type(dist, a, b) = result_type(dist, typeof(first(a)), typeof(first(b)))
```

Comment:
Should we do some `Base.IteratorEltype`-magic, depending on whether it returns `HasEltype` or `EltypeUnknown`? It would be cool to avoid reading out the `first` elements, particularly for empty iterables.

Reply:
Hmmm, will it be problematic to also use `eltype` here? Otherwise, it seems to be clearer to use

```julia
result_type(dist, a, b) = result_type(dist, first(a), first(b))
```

Reply:
Sounds like a good idea.

```julia
result_type(dist, a, b) = _result_type(dist, _eltype(a), _eltype(b))
_eltype(a) = __eltype(Base.IteratorEltype(a), a)
__eltype(::Base.HasEltype, a) = eltype(a)
__eltype(::Base.EltypeUnknown, a) = typeof(first(a))
```

Something like this, maybe.

Reply:
Yeah, that's cool. That let's `hamming("", "")` work out of the box. Nondegenerate cases were no problem before.